File: Basic data structure/javascript/stack_queue.js
// stack implementation

class Stack {
    constructor() {
        this.data = [];
    }
    push(item) {
        this.data.push(item);
    }

    pop() {
        return this.data.pop();
    }

    peek() {
        return this.data[this.data.length - 1];
    }
}

// queue implementation

class Queue {
    constructor() {
        this.data = [];
    }

    enqueue(item) {
        this.data.push(item);
    }

    dequeue() {
        return this.data.shift();
    }

    peek() {
        return this.data[0];
    }
}


stack = new Stack();
stack.push(1);
stack.push(2);
stack.push(3);
console.log(stack.peek());
console.log(stack.pop());
console.log(stack.pop());
console.log(stack.pop());
queue = new Queue();
queue.enqueue(1);
queue.enqueue(2);
queue.enqueue(3);
console.log(queue.peek());
console.log(queue.dequeue());
console.log(queue.dequeue());
console.log(queue.dequeue());

File: Basic data structure/go/stack_queue.go
// stack implementation
package main

import (
	"errors"
	"fmt"
)

type Stack struct {
	elements []interface{}
}

func (s *Stack) Push(e interface{}) {
	s.elements = append(s.elements, e)
}

func (s *Stack) Pop() (interface{}, error) {
	if len(s.elements) == 0 {
		return nil, errors.New("Empty stack!")
	}
	e := s.elements[len(s.elements)-1]
	s.elements = s.elements[:len(s.elements)-1]
	return e, nil
}

// check if Empty
func (s *Stack) isEmpty() bool {
	return len(s.elements) == 0
}

// queue
type Queue struct {
	elements []interface{}
}

func (q *Queue) enequeue(e interface{}) {
	q.elements = append(q.elements, e)
}

func (q *Queue) dequeue() (interface{}, error) {
	if len(q.elements) == 0 {
		return nil, errors.New("Empty queue!")
	}
	e := q.elements[0]
	q.elements = q.elements[1:]
	return e, nil
}

// check if Empty
func (q *Queue) isEmpty() bool {
	return len(q.elements) == 0
}

func main() {
	// For stack
	s := new(Stack)
	var stackElements []interface{}
	s.Push(1)
	s.Push(2)
	s.Push(3)
	s.Push(4)
	s.Push(5)
	for !s.isEmpty() {
		e, _ := s.Pop()
		stackElements = append(stackElements, e)
	}
	fmt.Println("Stack elements:", stackElements)

	// For queue
	q := new(Queue)
	var queueElements []interface{}
	q.enequeue(1)
	q.enequeue(2)
	q.enequeue(3)
	q.enequeue(4)
	q.enequeue(5)
	for !q.isEmpty() {
		e, _ := q.dequeue()
		queueElements = append(queueElements, e)
	}
	fmt.Println("Queue elements:", queueElements)

}
